### PR TITLE
Covid test screen

### PR DESCRIPTION
**Pull request by JoshElgar**
# Description
This PR adds several new questions to the Covid test screen (including swedish translations).
The new questions:
- ever_had_covid_test - boolean
- had_new_covid_test - boolean
- covid_test_result.- either positive, negative or test_failed
- covid_test_result_status - either waiting, received or never_had_test
- date_test_occurred (datetimefield) - either this or date_test_occurred_guess will be filled in.
- date_test_occurred_guess (choices)
- covid_test_mechanism (suggested choices but open field if they specify other)

There are two 'flows' through this page - users who are waiting for a test result, and users who are not.
- Users who are waiting for a test result should see q1: 'you recently told us you were waiting for a test result.. have you received those results?' If they respond yes (which corresponds to 'received'), they are then asked what the result was.

- Users who are not waiting for a test result are asked if they have ever had a test, or if they have had a new test since the last one. If they respond yes, they are asked to fill several questions around the test, when it occurred, how it was carried out etc. They are also asked if they've received results, to which they respond no (which corresponds to 'waiting'), and continue, or respond yes ('received') and are then asked what the result was.

- Users who mistakenly told us they did a test before, can break out of the 'waiting' loop by responding 'never had a test' to the question about having received results.

Note: Right now, the datepicker resets the date picked to the start of the day, and then converts it to UTC which means selecting '27/08/2020' actually corresponds in the backend to 26/08/2020 23:00pm.

## On what platform have you tested the change?

- [ ] Android - Emulator
- [X] Android - Physical device
- [ ] iOS - Emulator
- [ ] iOS - Physical device

## Screenshots
Screenshot_20200429-171052_2
Screenshot_20200429-171048_2
Screenshot_20200429-171036_2
Screenshot_20200429-171032_2
Screenshot_20200429-171022_2

## Out of scope and potential follow-up
Localisation of date picker..
Render question-by-question rather than all at once.

### CODE REVIEW

**Comment by MikeD-ZOE** [comment]
@Josh, do the dropdown items work properly on Android? I'm worried that because there's no Android-default given, the dropdown shows the first item, but this isn't what's set as the default/initial value for the field. So it looks to the user that the shown item is the selected one, but to the screen nothing has been selected. How is this being handled by the screen without a "Choose an option" first item in the dropdowns?

File: src/features/assessment/CovidTestScreen.tsx [hunk 1]
```diff
@@ -52,50 +67,59 @@ export default class CovidTestScreen extends Component<CovidProps, State> {
   }
 
   registerSchema = Yup.object().shape({
-    hasCovidTest: Yup.string().required(),
-    hasCovidPositive: Yup.string().required(),
-    takesAnyBloodPressureMedications: Yup.string(),
+    everHadCovidTest: Yup.string(),
+    covidTestResult: Yup.string(),
+    covidTestResultStatus: Yup.string(),
```
Comment (the-kenneth) [inline src/features/assessment/CovidTestScreen.tsx:72]: Fields can be left blank after pressing yes ive had a test. They would need some conditional required()

File: src/core/patient/PatientState.ts [hunk 2]
```diff
@@ -17,6 +17,8 @@ export type PatientStateType = {
   isSameHousehold: boolean;
   shouldAskLevelOfIsolation: boolean;
   shouldAskStudy: boolean;
+  isWaitingForCovidTestResult: boolean;
+  everHadCovidTest: boolean;
```
Comment (the-kenneth) [inline src/core/patient/PatientState.ts:21]: You havent added these to initial state, is there a reason for not defaulting these?

**Comment by JoshElgar** [comment]
> @josh, do the dropdown items work properly on Android? I'm worried that because there's no Android-default given, the dropdown shows the first item, but this isn't what's set as the default/initial value for the field. So it looks to the user that the shown item is the selected one, but to the screen nothing has been selected. How is this being handled by the screen without a "Choose an option" first item in the dropdowns?

Is the quick fix for this just to add an android default?

File: src/features/assessment/CovidTestScreen.tsx (hunk 1 above)
Comment (JoshElgar) [reply src/features/assessment/CovidTestScreen.tsx:72]: Yeah, had a feeling I would need to add a bit of validation.

File: src/core/user/UserService.ts [hunk 3]
```diff
@@ -198,6 +198,10 @@ export default class UserService extends ApiClientBase {
     const consent = await this.getConsentSigned();
     const shouldAskStudy = (isUSCountry() && consent && consent.document === 'US Nurses') || isGBCountry();
 
+    // CovidTestScreen flag
+    const isWaitingForCovidTestResult = patient.is_waiting_for_covid_test_result || false;
```
Comment (the-kenneth) [inline src/core/user/UserService.ts:202]: I cant see the field `is_waiting_for_covid_test_result ` getting updated from the client, so then in turn isWaitingForCovidTestResult is not updated?

File: src/core/patient/PatientState.ts (hunk 2 above)
Comment (JoshElgar) [reply src/core/patient/PatientState.ts:21]: As far as I understood it PatientState is more of a global state intended to represent the Patient as it passes through the screens. These values do actually get defaulted in updatePatientState in UserService.

File: src/core/user/UserService.ts (hunk 3 above)
Comment (JoshElgar) [reply src/core/user/UserService.ts:202]: Sorry, I should have been more clear about this. There is no `is_waiting_for_covid_test_result` field in the DB. There is a field called `covid_test_result_status` which stores the value 'waiting'. There is then a custom serializer field on Patient `is_waiting_for_covid_test_result` which checks if the latest assessment's `covid_test_result_status` is 'waiting'.

Comment (JoshElgar) [reply src/core/user/UserService.ts:202]: So even though it doesn't look like I'm updating `is_waiting_for_covid_test_result` directly, it is ( or should be, please test :) )actually updating.

**Comment by MikeD-ZOE** [comment]
> > @josh, do the dropdown items work properly on Android? I'm worried that because there's no Android-default given, the dropdown shows the first item, but this isn't what's set as the default/initial value for the field. So it looks to the user that the shown item is the selected one, but to the screen nothing has been selected. How is this being handled by the screen without a "Choose an option" first item in the dropdowns?
> 
> Is the quick fix for this just to add an android default?

Well, one of our open source contributors has this interesting way:

https://github.com/zoe/covid-tracker-react-native/blob/develop/src/features/patient/YourWorkScreen/index.tsx#L148

note the `.filter(Boolean)` at the end -- otherwise iOS collapses in a heap.

File: src/features/assessment/CovidTestScreen.tsx [hunk 4]
```diff
@@ -116,11 +141,52 @@ export default class CovidTestScreen extends Component<CovidProps, State> {
...
+      label: i18n.t('choose-one-of-these-options'),
+      value: '',
+    };
+
+    const dateToday = moment().toDate();
+
+    const covidTestResultItems = [
+      androidOption,
+      { label: 'No', value: 'negative' },
+      { label: 'Yes', value: 'positive' },
```
Comment (MikeD-ZOE) [inline src/features/assessment/CovidTestScreen.tsx:156]: These need to be i18n strings - this screen is used in Sweden

File: src/features/assessment/CovidTestScreen.tsx [hunk 5]
```diff
@@ -116,11 +141,52 @@ export default class CovidTestScreen extends Component<CovidProps, State> {
...
+      { label: i18n.t('covid-test.picker-test-failed'), value: 'test_failed' },
+    ].filter(Boolean) as IOption[];
+
+    const covidTestResultStatusItems = [
+      androidOption,
+      { label: 'Yes', value: 'received' },
+      { label: 'No', value: 'waiting' },
```
Comment (MikeD-ZOE) [inline src/features/assessment/CovidTestScreen.tsx:163]: These need to be i18n strings - this screen is used in Sweden

**Comment by MikeD-ZOE** [comment]
Screenshot 2020-04-29 at 18 23 50

> Warning: DatePickerIOS has been merged with DatePickerAndroid and will be removed in a future release. It can now be installed and imported from '@react-native-community/datetimepicker' instead of 'react-native'. See https://github.com/react-native-community/react-native-datetimepicker

**Comment by JoshElgar** [comment]
> Screenshot 2020-04-29 at 18 23 50
> 
> > Warning: DatePickerIOS has been merged with DatePickerAndroid and will be removed in a future release. It can now be installed and imported from '@react-native-community/datetimepicker' instead of 'react-native'. See https://github.com/react-native-community/react-native-datetimepicker

I tried using the react-native-datetimepicker directly but it was causing issues. I think this is because Native Base hasn't upgraded yet to use that yet.

File: src/core/user/UserService.ts (hunk 3 above)
Comment (the-kenneth) [reply src/core/user/UserService.ts:202]: When testing it doesnt show the isWaitingForCovidTestResult questions even though I indicated in the last assignment i was waiting

Comment (JoshElgar) [reply src/core/user/UserService.ts:202]: Sorry, fixed that on the backend now. I would hold off on further reviewing until I fix this validation as it's the last thing really.

File: src/features/assessment/CovidTestScreen.tsx [hunk 6]
```diff
@@ -92,11 +136,91 @@ export default class CovidTestScreen extends Component<CovidProps, State> {
...
+      everHadCovidTest: Yup.string(),
+      hadNewCovidTest: Yup.string(),
+      knowsDateOfTest: Yup.string().when(['hadNewCovidTest', 'everHadCovidTest'], {
+        is: (hadNewTest, everHadTest) => {
+          if (isWaitingForCovidTestResult) return false;
+          return hadNewTest === 'yes' || everHadTest === 'yes';
+        },
+        then: Yup.string().required(),
+      }),
+      dateTestOccurredGuess: Yup.mixed().when(['knowsDateOfTest', 'hadNewCovidTest', 'everHadCovidTest'], {
```
Comment (selvach) [inline src/features/assessment/CovidTestScreen.tsx:151]: Yup.mixed()?

File: src/features/assessment/CovidTestScreen.tsx (hunk 1 above)
Comment (JoshElgar) [reply src/features/assessment/CovidTestScreen.tsx]: Added validation now @the-kenneth

File: src/features/assessment/CovidTestScreen.tsx (hunk 6 above)
Comment (JoshElgar) [reply src/features/assessment/CovidTestScreen.tsx]: Fixed!

**Comment by JoshElgar** [comment]
Just a note: to test the ' waiting' flow locally, you will need my covid_test_questions branch of covid-app until it is in stage.

**Comment by selvach** [comment]
Could you update the stubs, if there's a new endpoint. Just keep it simple for now. Roberta was looking at more complex backend stubs?

**Comment by selvach** [comment]
And seeing as you are closest to merging a PR - could you `npm run lint:fix`

**Pull request merged**